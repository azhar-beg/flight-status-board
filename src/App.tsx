import React, { useState } from 'react';
import { createGlobalStyle, ThemeProvider } from 'styled-components';
import { Routes, Route } from 'react-router-dom';
import Flights from './components/Flights/Flights';
import FlightDetails from './components/FlightDetails/FlightDetails';
import Header from './components/Header/Header';
import { lightTheme, darkTheme } from './utils/theme';
import ErrorBoundary from './components/ErrorBoundary/ErrorBoundary';
import PageNotFound from './components/PageNotFound/PageNotFound';

const GlobalStyles = createGlobalStyle`
  html, body {
    height: 100%;
    overflow-x: hidden;
  }

  body {
    background-color: ${({ theme }) => theme.colors.background}; // Set background color from theme
  }
`;

function App() {
  const [isLightTheme, setIsLightTheme] = useState(true);
  const toggleTheme = () => {
    setIsLightTheme((prevTheme) => !prevTheme);
  };
  return (
    <ThemeProvider theme={isLightTheme ? lightTheme : darkTheme}>
      <GlobalStyles />
      <div>
        <Header toggleTheme={toggleTheme} isLightTheme={isLightTheme} />
        <ErrorBoundary>
          <Routes>
            <Route path="/" element={<Flights />} />
            <Route path="/flight-details/:id" element={<FlightDetails />} />
            <Route path="*" element={<PageNotFound />} />
          </Routes>
        </ErrorBoundary>
      </div>
    </ThemeProvider>
  );
}

export default App;
