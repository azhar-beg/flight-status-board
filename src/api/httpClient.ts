import axios from 'axios';

export const flightServiceClient = axios.create({
  baseURL: 'https://flight-status-mock.core.travelopia.cloud',
  timeout: 3000
});
