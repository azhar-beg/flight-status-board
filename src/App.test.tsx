import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { flightServiceClient } from './api/httpClient';
import App from './App';
import { MemoryRouter } from 'react-router-dom';

const mockedAxios = flightServiceClient as jest.Mocked<typeof flightServiceClient>;
jest.mock('./api/httpClient');

const mockFlightDetails = {
  id: 1,
  flightNumber: 'AA123',
  airline: 'American Airlines',
  origin: 'JFK',
  destination: 'LAX',
  departureTime: '2023-10-01T10:00:00Z',
  status: 'On Time'
};

describe('App Component', () => {
  beforeEach(() => {
    jest.spyOn(global.Math, 'random').mockReturnValue(0.123456789);
  });

  afterEach(() => {
    jest.spyOn(global.Math, 'random').mockRestore();
  })

  it('renders the Flight Status Board link', () => {
    render(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );
    const linkElement = screen.getByText(/Flight Status Board/i);
    expect(linkElement).toBeInTheDocument();
  });

  it('renders FlightDetails component for flight-details path', async () => {
    mockedAxios.get.mockResolvedValueOnce({ data: mockFlightDetails });
    render(
      <MemoryRouter initialEntries={['/flight-details/1']}>
        <App />
      </MemoryRouter>
    );
    await waitFor(() => expect(screen.getByText(/Flight Details/i)).toBeInTheDocument());
  });

  it('renders PageNotFound component for unknown path', () => {
    render(
      <MemoryRouter initialEntries={['/unknown']}>
        <App />
      </MemoryRouter>
    );
    expect(screen.getByText(/Page Not Found/i)).toBeInTheDocument();
  });

  it('should match previous snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
