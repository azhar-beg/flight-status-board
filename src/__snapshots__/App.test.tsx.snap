// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should match previous snapshot 1`] = `
<DocumentFragment>
  <div>
    <header
      class="sc-eOzmre eJRueW"
    >
      <div
        class="header-container"
      >
        <h1
          class="sc-eauhAA laKdas"
        >
          Flight Status Board
        </h1>
        <nav
          class="sc-fFoeYl fntnNB"
        >
          <ul>
            <li>
              <button
                class="sc-gQaihK chmoUq"
                id="home"
              >
                Home
              </button>
            </li>
          </ul>
          <label
            class="sc-iuUfFv kTzNpg"
          >
            <span>
              Light
            </span>
            <input
              class="sc-eUlrpB cLvERT"
              type="checkbox"
            />
            <div
              class="sc-dprtRQ gsdxzh"
            />
          </label>
        </nav>
      </div>
    </header>
    <div
      id="loader"
      style="display: flex; justify-content: center; align-items: center;"
    >
      <div
        class="sc-kLhKbu dBpLNs"
      >
        <img
          alt="Loading..."
          class="sc-ixGGxD imZNBa"
          src="loader.gif"
        />
        <p
          class="sc-khLCKb fEMHbM"
        >
          Remember that happiness is a way of travel, not a destination.
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
